src/common.ml

```diff
@@ -20,17 +20,16 @@ type token =
 
 (* Constants *)
 type const =
-      AtomConst of string    (* atoms    *)
-    | IntConst of int        (* integers *)
-    | FloatConst of float    (* floats   *)
-    | StringConst of string  (* strings  *)
-    | BoolConst of bool      (* booleans *)
+     Term of string * const list    (* atoms and compound terms *)
+   | Variable of string             (* vatiables *)     
+   | BoolConst of bool              (* booleans *)
 
 (* Expressions *)
 type exp =
-      VarExp of string              (* variables                *)
-    | ConstExp of const             (* constants                *)
-    | RuleExp of exp * exp          (* Head :- Body.            *)
-    | FactExp of exp                (* Head.                    *)
-    | CompoundTermExp of exp * exp  (* functor(arg1, arg2, ...) *)
-    | ListExp of exp * exp          (* arg1, arg2               *)
```

Comment:
I'm not sure if I understand any of the changes you made to `exp`.  Can you explain them? It seems like you moved most of the types of expressions to constants. You also currently have no way of representing a rule with conjunctions with this new AST.